### PR TITLE
Improve WithShadow

### PR DESCRIPTION
**Pull request by reaperrr**
Makes the trait public, upgradable and the offset customizable.

Uses the customizable offset to tweak shadow position of TD and RA aircraft, which looks a little better especially for landed aircraft (they no longer completely hide the shadow).

### CODE REVIEW

**Comment by Mailaender** [comment]
The improvement seems to be very subtle. :white_check_mark:

**Comment by reaperrr** [comment]
It's a little more visible with pixel-doubling.

File: OpenRA.Mods.Common/Traits/Render/WithShadow.cs [hunk 1]
```diff
@@ -16,30 +16,40 @@
...
+
+			if (self.IsDead || !self.IsInWorld)
+				return Enumerable.Empty<IRenderable>();
+
 			// Contrails shouldn't cast shadows
 			var height = self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length;
 			var shadowSprites = r.Where(s => !s.IsDecoration)
 				.Select(a => a.WithPalette(wr.Palette(info.Palette))
-					.OffsetBy(new WVec(0, 0, -height))
+					.OffsetBy(new WVec(info.Offset.X, info.Offset.Y, info.Offset.Z - height))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithShadow.cs]: You can simplify this to `info.Offset - new WVec(0, 0, height)`.

File: mods/cnc/rules/defaults.yaml [hunk 2]
```diff
@@ -153,6 +153,7 @@
 		GenericName: Helicopter
 	WithFacingSpriteBody:
 	WithShadow:
+		Offset: 43, 43, -85
```
Comment (pchote) [inline mods/cnc/rules/defaults.yaml]: Putting the shadows below ground level isn't great and may cause rendering glitches as we progress further with zbuffer support.  Can you move that offset into Y instead?

**Comment by reaperrr** [comment]
Don't merge this yet, one more fixup coming in.

**Comment by reaperrr** [comment]
Updated. ZOffset is now customziable as well.

That's necessary, because positive Y values otherwise make the shadow overlap the body sprite.

**Comment by GraionDilach** [comment]
Could WithShadow get an option that instead of the actor, it'd generate the sahdow from a custom overlay? RA2 used a "blob" shadow from jjshadow.shp instead of this method.

**Comment by reaperrr** [comment]
I don't feel like doing that via this PR.

Besides, you can already do that by adding a `WithSpriteBody@shadow` with `Sequence: shadow` to your actor (that's what I do in my MechCommander mod).

File: OpenRA.Mods.Common/Traits/Render/WithShadow.cs [hunk 3]
```diff
@@ -16,31 +16,44 @@
...
+	public class WithShadowInfo : UpgradableTraitInfo
 	{
 		[PaletteReference] public readonly string Palette = "shadow";
 
-		public object Create(ActorInitializer init) { return new WithShadow(this); }
+		[Desc("Shadow position offset relative to actor position (ground level).")]
+		public readonly WVec Offset = WVec.Zero;
+
+		[Desc("Shadow Z offset relative to actor sprite.")]
+		public readonly int ZOffset = -5;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/WithShadow.cs]: Hm, couldn't we just include this in the normal `Offset`?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithShadow.cs]: No, they are very different things.

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithShadow.cs]: > Hm, couldn't we just include this in the normal Offset?

`Offset` is a world position offset, while `ZOffset` could also be called `ZSortingOffset`, but we already call it just `ZOffset` everywhere else.

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/b7ced950406170fb5d2a95cb5081590884b8620e...686afa9caf1a5b05407d104fd56bf039e299d21b)

**Pull request merged**